test/fp32-limits.js

```diff
@@ -0,0 +1,74 @@
+import destination from '@turf/destination';
+import {lngLatToWorld, getDistanceScales} from 'viewport-mercator-project';
+import test from 'tape-catch';
+
+import VIEWPORT_PROPS from './utils/sample-viewports';
+
+function getDiff(value, baseValue) {
+  const errorPixels = value.map((v, i) => Math.abs(v - baseValue[i]));
+  const error = value.map(
+    (v, i) => Math.abs(v - baseValue[i]) / Math.min(Math.abs(v), Math.abs(baseValue[i]))
+  );
+
+  return {
+    errorPixels,
+    error,
+    message: `off by \
+      (${errorPixels.map(d => d.toFixed(3)).join(', ')}) pixels, \
+      (${error.map(d => `${(d * 100).toFixed(3)}%`).join(', ')})`
+  };
+}
+
+test('FP32 & Offset Comparison', t => {
+  // Explore limits at different scales
+  for (let zoom = 1; zoom <= 20; zoom++) {
+    const scale = Math.pow(2, zoom);
+    t.comment('--------');
+    t.comment(`Zoom = ${zoom}, Scale = ${scale}`);
+
+    for (const vc in VIEWPORT_PROPS) {
+      t.comment(vc);
+      const props = VIEWPORT_PROPS[vc];
+      const {longitude, latitude} = props;
+      const {pixelsPerMeter, pixelsPerMeter2} = getDistanceScales({
+        latitude,
+        longitude,
+        scale,
+        highPrecision: true
+      });
+
+      // Test distance from one edge of the viewport to the other
+      const delta = Math.pow(2, 20 - zoom) * 50;
```

Comment:
How is `50` picked?

Reply:
I estimate a distance that covers most of the viewport.
At zoom level 20, I think 50 meters cover a decent part of the viewport.
Then I scale it up based on the zoom level to cover the same part of screen.

Reply:
The viewport extent is dependent on width/height/latitude. You need to make this choice explicit.

```js
function getHalfViewportSizeInMeters(opts) {
  var v = new WebMercatorViewport(Object.assign({}, opts, {zoom:20}));
  var tl = v.unproject([0, 0]);
  var br = v.unproject([v.width / 2, v.height / 2]);
  var scales = getDistanceScales(v);
  return [
    (br[0] - tl[0]) * scales.metersPerPixel[0] * scales.pixelsPerDegree[0],
    (tl[1] - br[1]) * scales.metersPerPixel[1] * scales.pixelsPerDegree[1]
  ];
}

getHalfViewportSizeInMeters({width: 800, height: 600, longitude: 0, latitude: 0});
// [ 29.824674136112748, 22.36850559713332 ]
getHalfViewportSizeInMeters({width: 3200, height: 3200, longitude: 0, latitude: 0});
// [ 119.29869652464602, 119.29869652464602 ]
getHalfViewportSizeInMeters({width: 3200, height: 3200, longitude: 0, latitude: 70});
// [ 40.8025572839249, 40.8025572839249 ]

```

Reply:
Since the projection is really using lnglat offset instead of meter offset, you can skip the converting unprojected coordinates to meters

Reply:
But how much coord difference do I need to test?